frontend/src/component/pure_module.rs

```diff
@@ -0,0 +1,42 @@
+use std::mem;
+use yew::{Component, ComponentLink, Html, ShouldRender, Properties};
+use mixlab_protocol::ModuleId;
+use crate::workspace::Window;
+
+pub trait PureModule: Clone + PartialEq + 'static {
+    fn view(&self, id: ModuleId, module: ComponentLink<Window>) -> Html;
+}
+
+#[derive(Properties, Clone)]
+pub struct PureProps<Params: PureModule> {
+    pub id: ModuleId,
+    pub params: Params,
+    pub module: ComponentLink<Window>,
+}
+
+pub struct Pure<Params: PureModule> {
+    props: PureProps<Params>
+}
+
+impl<Params: PureModule> Component for Pure<Params> {
+    type Properties = PureProps<Params>;
+    type Message = ();
+
+    fn create(props: Self::Properties, _: ComponentLink<Self>) -> Self {
+        Self { props }
+    }
+
+    fn update(&mut self, _msg: Self::Message) -> ShouldRender {
+        false
+    }
+
+    fn change(&mut self, mut props: Self::Properties) -> ShouldRender {
+        mem::swap(&mut self.props, &mut props);
+
+        self.props.id != props.id || self.props.params != props.params
```

Comment:
This should give us a nice little performance boost for free for all pure modules

Reply:
Haven't you just swapped `self.props` before this check?

Also, did you read https://yew.rs/docs/advanced-topics/optimizations#neq_assign?

Reply:
https://github.com/yewstack/yewtil

Has the `NeqAssign` trait and it also has a "pure components" trait out of the box, it seems.

Reply:
> Haven't you just swapped `self.props` before this check?

Oohhh `swap` not `set`. Never mind!

Reply:
I find it a bit odd that the change function would take new props mutably, though. Then the interface here is that after updating a component, the props in hand are now the old props (from caller's perspective)... 🤨

Reply:
>I find it a bit odd that the change function would take new props mutably

The `mut` in that place just means that the `props` binding is mutable inside the function. It's a funny bit of syntax but there's no visible difference to the prototype outside the function (otherwise we wouldn't be able to impl the trait!)

> https://github.com/yewstack/yewtil
> Has the NeqAssign trait and it also has a "pure components" trait out of the box, it seems.

Oh this is handy! I hadn't seen this. Unfortunately `ComponentLink` is not `PartialEq` so we can't use it here